doc/relevancy.md

```diff
@@ -0,0 +1,442 @@
+
+## Synonyms
+
+A synonym is a word or phrase that means exactly or nearly the same as another
+word or phrase in the same language, for example shut is a synonym of close.
+
+Synonyms are defined in the [synonyms.yml file](https://github.com/alphagov/rummager/blob/master/config/schema/synonyms.yml).
+
+Each synonym group is represented as comma separated lists of synonyms, and
+optionally a "=>" symbol.
+
+### Grouping of synonyms
+
+The use of => is known as _Simple contraction_. It maps a group of synonyms on
+the left side to a single value on the right side:
+
+"leap,hop => jump"
+
+means "a search for 'leap' or 'hop' should return documents with 'jump' in them.
+
+It must be applied both at index time and at query time, to ensure that query
+terms are mapped to the same single value that exists in the index.
+
+Comma separated "foo, bar, baz" means to return any of those terms.
+
+### Filtering
+
+Additional configuration is defined in the elasticsearch_schema.yml and
+stems.yml files.  This configuration is merged with the JSON configuration,
+and then passed to elasticsearch directly.
+
+Within elasticsearch_schema.yml, as well as an index_synonym filter there is
+also a synonym_protwords filter.
+
+This marks the terms generated from the synonym matching process as "keywords".
+This prevents them being stemmed, and is required if there is a stemming filter
+following the synonym filter in the analysis chain. This will make sure phrases
+stay together aka "heart attack".
+
+This blog post suggests that as well as using keywords, it can be useful to use
+'keep words' to only filter the phrases we want.
+
+The post suggests to first generate shingles, followed by synonyms, followed
+by keepwords.
+
+In other words: generate candidate keyphrases by shingling, expand them with
+synonyms, then cull out any non-synonyms with keepwords.
+
+### Old synonyms
+
+Synonyms are also specified within old_synonyms.yml which contains a set of
+synonyms which were applied in an old, and moderately broken, way.
+
+According to the comments:
+
+"This is kept separate from the `synonyms.yml` file so that we can support
+and modify the lists for the new synonym approach without changing things
+for the old approach.  Once we're happy with the new approach, we'll delete
+the `old_synonyms.yml` file, and remove support for using it." (Nov 2016)
+
+The analyzer which use the old_synonyms filter do not also use the
+synonym_protwords which could mean less effective results.
+
+Not too sure why we have not migrated the old_synonyms.
+
+### Summary
+
+Things to further investigate:
+
+- Are the synonymn groupings effective. Does simple contraction make much difference?
+- Currently with_index_synonyms filter looks like this: filter: [standard,
+asciifolding, lowercase, index_synonym, synonym_protwords, stemmer_override,
+stemmer_english]. Maybe we could try applying shingles, then synonyms and then
+applying keepwords (like suggested in the blog post)
```

Comment:
This is inaccurate, but the shingle / keepwords suggestion could still be something to try.